src/Domain/Python/FunctionsOfTime.cpp

```diff
@@ -21,22 +23,99 @@ namespace domain::py_bindings {
-  py::class_<FunctionsOfTime::FunctionOfTime>(m, "FunctionOfTime")
+  py::class_<FunctionsOfTime::FunctionOfTime,
+             std::shared_ptr<FunctionsOfTime::FunctionOfTime>>(m,
+                                                               "FunctionOfTime")
       .def("time_bounds", &FunctionsOfTime::FunctionOfTime::time_bounds)
       .def("func", &FunctionsOfTime::FunctionOfTime::func, py::arg("t"))
       .def("func_and_deriv", &FunctionsOfTime::FunctionOfTime::func_and_deriv,
            py::arg("t"))
       .def("func_and_2_derivs",
            &FunctionsOfTime::FunctionOfTime::func_and_2_derivs, py::arg("t"));
+  m.def(
+      "serialize_functions_of_time",
+      [](const std::unordered_map<
+          std::string,
+          std::shared_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+             shared_deserialized_functions_of_time) {
+        std::unordered_map<
+            std::string,
+            std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+            unique_deserialized_functions_of_time{};
+        if (shared_deserialized_functions_of_time.contains("Expansion")) {
+          unique_deserialized_functions_of_time["Expansion"] =
+              shared_deserialized_functions_of_time.at("Expansion")
+                  ->get_clone();
+        }
+        if (shared_deserialized_functions_of_time.contains(
+                "ExpansionOuterBoundary")) {
+          unique_deserialized_functions_of_time["ExpansionOuterBoundary"] =
+              shared_deserialized_functions_of_time
+                  .at("ExpansionOuterBoundary")
+                  ->get_clone();
+        }
+        if (shared_deserialized_functions_of_time.contains("Rotation")) {
+          unique_deserialized_functions_of_time["Rotation"] =
+              shared_deserialized_functions_of_time.at("Rotation")->get_clone();
+        }
+        if (shared_deserialized_functions_of_time.contains("Translation")) {
+          unique_deserialized_functions_of_time["Translation"] =
+              shared_deserialized_functions_of_time.at("Translation")
+                  ->get_clone();
+        }
+        return serialize<std::unordered_map<
+            std::string,
+            std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>>(
+            unique_deserialized_functions_of_time);
+      },
+      py::arg("deserialized_functions_of_time"));
   m.def(
       "deserialize_functions_of_time",
       [](const std::vector<char>& serialized_functions_of_time) {
         return deserialize<std::unordered_map<
             std::string,
-            std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>>(
+            std::shared_ptr<domain::FunctionsOfTime::FunctionOfTime>>>(
             serialized_functions_of_time.data());
       },
       py::arg("serialized_functions_of_time"));
+  py::class_<FunctionsOfTime::PiecewisePolynomial<3>,
+             FunctionsOfTime::FunctionOfTime,
+             std::shared_ptr<FunctionsOfTime::PiecewisePolynomial<3>>>(
```

Comment:
For the PiecewisePolynomial, let's bind both order 2 and 3. To avoid duplicating code, write a lambda that's templated and call that for 2 and 3 like so

```cpp
const auto bind_piecewise_polynomial = [&]<size_t Order>() {
  // code that you have now using Order
  // Have each name be "PiecewisePolynomial" + std::to_string(Order)
};

bind_piecewise_polynomial.template operator()<2>();
bind_piecewise_polynomial.template operator()<3>();
```

(no need to do this for the Quat)